flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/expressions/analysis/FieldBacktracker.scala

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.table.expressions.analysis
+
+import org.apache.flink.api.table.expressions.{Naming, ResolvedFieldReference}
+import org.apache.flink.api.table.input.{AdaptiveTableSource, TableSource}
+import org.apache.flink.api.table.plan._
+
+object FieldBacktracker {
+
+  /**
+   * Tracks a field back to its Root and returns its original name and AdaptiveTableSource
+   * if possible.
+   * This only happens if the field is forwarded unmodified. Renaming operations are reverted.
+   *
+   * @param op start operator
+   * @param fieldName field name at start operator
+   * @return original field name with corresponding AdaptiveTableSource
+   */
+  def resolveFieldNameAndTableSource(op: PlanNode, fieldName: String):
+      Option[(AdaptiveTableSource, String)] = {
+    op match {
+      case s@Select(input, selection) =>
+        var resolvedField: Option[(AdaptiveTableSource, String)] = None
+        // only follow unmodified fields
+        selection.foreach(expr => {
+          if (resolvedField.isEmpty) {
+            expr match {
+              case ResolvedFieldReference(name, _) if name == fieldName =>
+                resolvedField = resolveFieldNameAndTableSource(input, fieldName)
+              case n@Naming(ResolvedFieldReference(oldName, _), newName) if newName == fieldName =>
+                resolvedField = resolveFieldNameAndTableSource(input, oldName)
+              case _ => // do nothing
+            }
+          }
+        })
```

Comment:
But here again you will traverse the whole `selection` sequence. Is there any special reason for it? Why not simply calculate the elements which are needed to find the first resolved field? With Scala you have all the possibilities at hand and furthermore you get rid of the `var` variable:

```
val resolvedField = selection.view.flatMap{
  case ResolvedFieldReference(name, _) if name == fieldName =>
    resolvedField = resolveFieldNameAndTableSource(input, fieldName)
  case n@Naming(ResolvedFieldReference(oldName, _), newName) if newName == fieldName =>
    resolvedField = resolveFieldNameAndTableSource(input, oldName)
  case _ => // do nothing
}.headOption
```